clients/drcachesim/simulator/cache_lru.cpp

```diff
@@ -69,7 +69,5 @@ cache_lru_t::replace_which_way(int line_idx)
             max_way = way;
         }
     }
-    // Set to non-zero for later access_update optimization on repeated access
-    get_caching_device_block(line_idx, max_way).counter_ = 1;
```

Comment:
See my analysis at https://github.com/DynamoRIO/dynamorio/issues/4881#issuecomment-827660933: removing this line will cause the ==0 short-circuit to always be executed, resulting in all the counters permanently staying at 0, right?

I would vote for my suggestion in the linked comment of initializing the counters to 0..assoc-1.

Reply:
All the counters staying at 0? Why? Considering a 4-line set. When a line comes in, all counters are 0, 0 <= 0, every counter will ++, it will become:

- a 0
- invalid 1
- invalid 1
- invalid 1

You can imagine these lines to be like a linkded-list. When visiting a node, (it will be removed from the middle and inserted at head), every node before it will be put behind a position(index++, if it has a index)(similar to all counter++ which has value <= cnt). And then the new head index becomes 0. Invalid lines counter will always be the max value, we can ignore it.

So where do I make a mistake? In my opinion any initializing value is OK... Could you give me a counter-example?

Reply:
All the counters staying at 0? Why? Considering a 4-line set. When a line comes in, all counters are 0, 0 <= 0, every counter will ++, it will become:

- a 0
- invalid 1
- invalid 1
- invalid 1

You can imagine these lines to be like a linkded-list. When visiting a node, (it will be removed from the middle and inserted at head), every node before it will be put behind a position(index++, if it has a index)(similar to all counter++ which has value <= cnt). And then the new head index becomes 0. Invalid lines counter will always be the max value, we can ignore it.

So where do I make a mistake? In my opinion any initializing value is OK... Could you give me a counter-example?

Reply:
See cache_lru.cpp lines 44-46:
```C
    // Optimization: return early if it is a repeated access.
    if (cnt == 0)
        return;
```
Counters are all initialized to 0 and this `cnt == 0` branch will be taken every time.  Thus no counter will ever be incremented.

Reply:
oh my fault, I ignore this . Sorry

Reply:
Can we set the initial counter value to 1? I think it can also work. Changing it to 0,1,..associativity - 1 looks a bit complex by changing code.

Reply:
> Can we set the initial counter value to 1? I think it can also work. 

Intuitively the code seems clearest to me to start out in the steady state of 0..assoc-1 and then there is no separate bootstrapping stage and it is just easier to think about.

> Changing it to 0,1,..associativity - 1 looks a bit complex by changing code.

It should be the same amount of code as changing it to 1: just override `init` and have a very simple loop there.  E.g., see `cache_fifo_t::init`.

Reply:
Yep. Thanks for guidance. I will push new code later.